File: python/tests/test_session_fs.py
"""Test Meterpreter filesystem operations."""

import pytest


@pytest.mark.asyncio
async def test_session_fs_operations(client):
    """Test Meterpreter filesystem operations.

    Note: This test requires an active Meterpreter session.
    If no sessions exist, the test will be skipped.
    """
    # Get active sessions
    session_ids = await client.list_sessions()

    if not session_ids:
        pytest.skip("No active sessions available for FS operations test")

    # Use first session
    session_id = session_ids[0]

    # Get session info to check if it's Meterpreter
    session_info = await client.session_get(session_id)
    session_type = session_info.get("type", "")

    if "meterpreter" not in session_type.lower():
        pytest.skip(f"Session {session_id} is not Meterpreter (type: {session_type})")

    print(f"\n✓ Testing with Meterpreter session {session_id} (type: {session_type})")

    # Test 1: Get current working directory
    pwd = await client.session_fs_pwd(session_id)
    assert pwd, "PWD should not be empty"
    print(f"  Current directory: {pwd}")

    # Test 2: Get path separator
    separator = await client.session_fs_separator(session_id)
    assert separator in ["/", "\\"], f"Invalid separator: {separator}"
    print(f"  Path separator: {repr(separator)}")

    # Test 3: List current directory
    entries = await client.session_fs_ls(session_id, pwd)
    assert isinstance(entries, list), "Entries should be a list"
    print(f"  Directory has {len(entries)} entries")
    if entries:
        print(f"    First few: {entries[:5]}")

    # Test 4: Check if a common path exists
    test_path = "C:\\Windows\\System32" if separator == "\\" else "/tmp"
    exists = await client.session_fs_exists(session_id, test_path)
    print(f"  Path '{test_path}' exists: {exists}")

    # Test 5: Get file stats
    if exists:
        stat = await client.session_fs_stat(session_id, test_path)
        assert isinstance(stat, dict), "Stat should return a dict"
        print(f"  Stat info: ftype={stat.get('ftype')}, is_directory={stat.get('is_directory')}")

    # Test 6: Expand path (test environment variables)
    if separator == "\\":
        test_expand = "%TEMP%"
    else:
        test_expand = "$HOME"

    try:
        expanded = await client.session_fs_expand_path(session_id, test_expand)
        print(f"  Expanded '{test_expand}' to: {expanded}")
        assert test_expand not in expanded, "Path should be expanded"
    except Exception as e:
        print(f"  Warning: expand_path failed: {e}")

    print("✓ Meterpreter FS read operations work correctly")


@pytest.mark.asyncio
async def test_session_fs_file_operations(client):
    """Test Meterpreter file creation/deletion operations.

    Note: This test requires an active Meterpreter session.
    If no sessions exist, the test will be skipped.
    """
    # Get active sessions
    session_ids = await client.list_sessions()

    if not session_ids:
        pytest.skip("No active sessions available for FS operations test")

    # Use first session
    session_id = session_ids[0]

    # Get session info to check if it's Meterpreter
    session_info = await client.session_get(session_id)
    session_type = session_info.get("type", "")

    if "meterpreter" not in session_type.lower():
        pytest.skip(f"Session {session_id} is not Meterpreter (type: {session_type})")

    print(f"\n✓ Testing file operations with session {session_id}")

    # Get separator to construct proper paths
    separator = await client.session_fs_separator(session_id)
    pwd = await client.session_fs_pwd(session_id)

    # Create test directory path
    test_dir = f"{pwd}{separator}test_assassinate_dir"
    test_file = f"{test_dir}{separator}test_file.txt"

    try:
        # Test 1: Create directory
        print(f"  Creating directory: {test_dir}")
        await client.session_fs_mkdir(session_id, test_dir)

        # Verify directory was created
        exists = await client.session_fs_exists(session_id, test_dir)
        assert exists, f"Directory {test_dir} should exist after creation"
        print("  ✓ Directory created successfully")

        # Test 2: List directory (should be empty)
        entries = await client.session_fs_ls(session_id, test_dir)
        # Might have . and .. entries
        print(f"  Directory has {len(entries)} entries: {entries}")

        # Test 3: Remove directory
        print(f"  Removing directory: {test_dir}")
        await client.session_fs_rmdir(session_id, test_dir)

        # Verify directory was removed
        exists = await client.session_fs_exists(session_id, test_dir)
        assert not exists, f"Directory {test_dir} should not exist after removal"
        print("  ✓ Directory removed successfully")

    except Exception as e:
        # Clean up on error
        print(f"  Error during test: {e}")
        try:
            # Try to remove test directory if it exists
            if await client.session_fs_exists(session_id, test_dir):
                await client.session_fs_rmdir(session_id, test_dir)
        except:
            pass
        raise

    print("✓ Meterpreter FS write operations work correctly")


@pytest.mark.asyncio
async def test_session_fs_no_session():
    """Test that FS operations fail gracefully when session doesn't exist."""
    from assassinate.ipc.client import MsfClient

    client = MsfClient()
    await client.connect()

    try:
        # Try to use FS operation on non-existent session
        with pytest.raises(Exception):
            await client.session_fs_pwd(99999)

        print("✓ FS operations properly fail for non-existent sessions")
    finally:
        await client.disconnect()

File: rust/bridge/tests/test_session_fs.rs
// Test: Meterpreter filesystem operations
// Requires an active Meterpreter session to fully test

mod common;

use bridge::ruby_bridge;
use magnus::{TryConvert, value::ReprValue};

#[test]
fn it_tests_meterpreter_fs_operations() {
    let (ruby, framework) = common::init_framework();

    // Get sessions
    let sessions = ruby_bridge::call_method(framework, "sessions", &[])
        .expect("Failed to get sessions");
    let session_keys = ruby_bridge::call_method(sessions, "keys", &[])
        .expect("Failed to get session keys");
    let session_ids: Vec<i64> = TryConvert::try_convert(session_keys)
        .unwrap_or_else(|_| Vec::new());

    if session_ids.is_empty() {
        println!("⚠ No active sessions - skipping Meterpreter FS test");
        println!("  To test FS operations, create a Meterpreter session first");
        return;
    }

    let session_id = session_ids[0];
    println!("Testing with session ID: {}", session_id);

    // Get session object
    let sid_val = ruby.integer_from_i64(session_id).as_value();
    let session_val = ruby_bridge::call_method(sessions, "[]", &[sid_val])
        .expect("Failed to get session");

    // Check session type
    let type_val = ruby_bridge::call_method(session_val, "type", &[])
        .expect("Failed to get type");
    let session_type = ruby_bridge::value_to_string(type_val)
        .expect("Failed to convert type");
    println!("Session type: {}", session_type);

    if !session_type.to_lowercase().contains("meterpreter") {
        println!("⚠ Session is not Meterpreter - skipping FS tests");
        return;
    }

    let session = bridge::Session::from_raw(session_val, session_id);

    // Test fs_pwd
    match session.fs_pwd() {
        Ok(pwd) => {
            println!("✓ fs_pwd() = {}", pwd);
            assert!(!pwd.is_empty(), "PWD should not be empty");
        }
        Err(e) => println!("⚠ fs_pwd() failed: {}", e),
    }

    // Test fs_separator
    match session.fs_separator() {
        Ok(sep) => {
            println!("✓ fs_separator() = {:?}", sep);
            assert!(sep == "/" || sep == "\\", "Separator should be / or \\");
        }
        Err(e) => println!("⚠ fs_separator() failed: {}", e),
    }

    // Test fs_ls on a safe path
    let test_path = if session_type.contains("windows") {
        "C:\\Windows"
    } else {
        "/tmp"
    };

    match session.fs_ls(test_path) {
        Ok(entries) => {
            println!("✓ fs_ls({}) returned {} entries", test_path, entries.len());
            if !entries.is_empty() {
                println!("  First few: {:?}", &entries[..entries.len().min(3)]);
            }
        }
        Err(e) => println!("⚠ fs_ls() failed: {}", e),
    }

    // Test fs_exists
    match session.fs_exists(test_path) {
        Ok(exists) => {
            println!("✓ fs_exists({}) = {}", test_path, exists);
            assert!(exists, "Test path should exist");
        }
        Err(e) => println!("⚠ fs_exists() failed: {}", e),
    }

    // Test fs_stat
    match session.fs_stat(test_path) {
        Ok(stat) => {
            println!("✓ fs_stat({}) returned stat info", test_path);
            if let Some(ftype) = stat.get("ftype") {
                println!("  ftype: {}", ftype);
            }
        }
        Err(e) => println!("⚠ fs_stat() failed: {}", e),
    }

    println!("✓ Meterpreter FS operations work correctly");
}
